web-components/src/components/tabs/Tabs.ts

```diff
@@ -254,39 +300,197 @@ export namespace Tabs {
         acc[tab.id] = idx;
         return acc;
       }, {} as Record<TabId, number>);
-
-      await this.manageOverflow();
     }
 
     get slotItem() {
       return this.tabSlotElement;
     }
 
     protected filterSlotted() {
-      return this.tabSlotElement!.assignedElements() as HTMLElement[];
+      return this.tabSlotElement.assignedElements() as HTMLElement[];
     }
 
     protected async handleResize(contentRect: DOMRect) {
       super.handleResize && super.handleResize(contentRect);
       await this.manageOverflow();
     }
 
-    private updateSelectedTab(newSelectedIndex: number) {
-      const { tabs, panels } = this;
+    private dispatchDragEvent(eventName: string, srcEvent: Sortable.SortableEvent | Sortable.MoveEvent) {
+      this.dispatchEvent(
+        new CustomEvent<{ srcEvent: Sortable.SortableEvent | Sortable.MoveEvent }>(eventName, {
+          composed: true,
+          bubbles: true,
+          detail: {
+            srcEvent
+          }
+        })
+      );
+    }
+
+    handleOnEnd = async (event: Sortable.SortableEvent) => {
+      event.stopPropagation();
+      const oldIndex = event.oldIndex;
+      const newIndex = event.newIndex;
+
+      const visibleTabElements = [...this.tabsFilteredAsVisibleList];
+      const hiddenTabElements = [...this.tabsFilteredAsHiddenList];
+
+      if (
+        event.to === this.visibleTabsContainerElement &&
+        event.from === this.hiddenTabsContainerElement &&
+        oldIndex !== undefined &&
+        newIndex !== undefined
+      ) {
+        this.visibleTabsContainerElement.querySelector("#" + event.item.id)?.remove();
+        const draggedElement = hiddenTabElements[oldIndex];
+        const autoMoveElement = visibleTabElements.slice(-1)[0];
+        hiddenTabElements.splice(oldIndex, 1);
+        hiddenTabElements.splice(0, 0, autoMoveElement);
+        const newDraggedIndex = newIndex === this.tabsFilteredAsVisibleList.length ? newIndex - 1 : newIndex;
+
+        visibleTabElements.splice(newDraggedIndex, 0, draggedElement);
+        visibleTabElements.pop();
+
+        this.tabsFilteredAsVisibleList = visibleTabElements;
+        this.tabsFilteredAsHiddenList = hiddenTabElements;
+      } else if (
+        event.from === this.visibleTabsContainerElement &&
+        event.to === this.hiddenTabsContainerElement &&
+        oldIndex !== undefined &&
+        newIndex !== undefined
+      ) {
+        this.hiddenTabsContainerElement.querySelector("#" + event.item.id)?.remove();
```

Comment:
extract method for readiblity, handleOnEnd to too big function, extract small functions for each conditions

Reply:
Sure will break it up